runtime/eventlog.c

```diff
@@ -0,0 +1,257 @@
+#define CAML_INTERNALS
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <inttypes.h>
+#include "caml/alloc.h"
+#include "caml/eventlog.h"
+#include "caml/osdeps.h"
+
+#define CTF_MAGIC 0xc1fc1fc1
+#define CAML_TRACE_VERSION 0x1
+
+struct ctf_stream_header {
+  uint32_t magic;
+  uint16_t caml_trace_version;
+  uint16_t stream_id;
+};
+
+static struct ctf_stream_header header = {
+  CTF_MAGIC,
+  CAML_TRACE_VERSION,
+  0
+};
+
+#pragma pack(1)
+struct ctf_event_header {
+  uint64_t timestamp;
+  uint32_t pid;
+  uint32_t id;
+};
+
+static uintnat alloc_buckets [20] =
+  {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
+
+
+struct event {
+  struct ctf_event_header header;
+  uint8_t  phase; /* for GC events */
+  uint8_t  counter_kind; /* misc counter name */
+  uint8_t  alloc_bucket; /* for alloc counters */
+  uint32_t count; /* for misc counters */
+};
+
+static FILE* output;
+static uintnat eventlog_startup_timestamp = 0;
+static uint32_t eventlog_startup_pid = 0;
+static uintnat eventlog_paused = 0;
```

Comment:
Should this be part of the domain state structure? Eventually, I would like each domain to be able to start and stop logging independently.

Reply:
I think it's a good idea,
Is it too much state to be added there? I don't think we'll require much more of these assuming we stay on a one stream per domain logic.

Reply:
This was done in d4f25e7